source/arch/encoding_simd_sse41.c

```diff
@@ -0,0 +1,396 @@
+        5, 4, 3, 0xFF,
+        8, 7, 6, 0xFF,
+        11, 10, 9, 0xFF,
+
+        2, 1, 0, 0xFF,
+        5, 4, 3, 0xFF,
+        8, 7, 6, 0xFF,
+        11, 10, 9, 0xFF
+        /* LSB */
+        /* clang-format on */
+    }};
+    vec = _mm256_shuffle_epi8(vec, _mm256_load_si256((__m256i const *)&shufvec_buf));
+
+    /*
+     * Now shift and mask to split out 6-bit groups.
+     * We'll also do a second byteswap to get back into big-endian
+     */
+    __m256i mask0 = _mm256_set1_epi32(0x3F);
+    __m256i mask1 = _mm256_set1_epi32(0x3F << 6);
+    __m256i mask2 = _mm256_set1_epi32(0x3F << 12);
+    __m256i mask3 = _mm256_set1_epi32(0x3F << 18);
+
+    __m256i digit0 = mask0 & vec;
+    __m256i digit1 = mask1 & vec;
+    __m256i digit2 = mask2 & vec;
+    __m256i digit3 = mask3 & vec;
+
+    /*
+     * Because we want to byteswap, the low-order digit0 goes into the
+     * high-order byte
+     */
+    digit0 = _mm256_slli_epi32(digit0, 24);
+    digit1 = _mm256_slli_epi32(digit1, 10);
+    digit2 = _mm256_srli_epi32(digit2, 4);
+    digit3 = _mm256_srli_epi32(digit3, 18);
+
+    vec = (digit0 | digit1) | (digit2 | digit3);
+
+    /* Finally translate to the base64 character set */
+    return encode_chars(vec);
+}
+
+void aws_common_private_base64_encode_sse41(const uint8_t *input, uint8_t *output, size_t inlen) {
+    __m256i instride, outstride;
+
+    while (inlen >= 32) {
+        /*
+         * Where possible, we'll load a full vector at a time and ignore the over-read.
+         * However, if we have < 32 bytes left, this would result in a potential read
+         * of unreadable pages, so we use bounce buffers below.
+         */
+        instride = _mm256_loadu_si256((__m256i const *)input);
+        outstride = encode_stride(instride);
+        _mm256_storeu_si256((__m256i *)output, outstride);
+
+        input += 24;
+        output += 32;
+        inlen -= 24;
+    }
+
+    while (inlen) {
+        /*
+         * We need to go through a bounce buffer for anything remaining, as we
+         * don't want to over-read or over-write the ends of the buffers.
+         */
+        size_t stridelen = inlen > 24 ? 24 : inlen;
+        size_t outlen = ((stridelen + 2) / 3) * 4;
+
+        memset(&instride, 0, sizeof(instride));
+        memcpy(&instride, input, stridelen);
+
+        outstride = encode_stride(instride);
+        memcpy(output, &outstride, outlen);
+
+        if (inlen < 24) {
+            if (inlen % 3 >= 1) {
+                /* AA== or AAA= */
+                output[outlen - 1] = '=';
+            }
+            if (inlen % 3 == 1) {
+                /* AA== */
+                output[outlen - 2] = '=';
+            }
+
+            return;
+        }
+
+        input += stridelen;
+        output += outlen;
+        inlen -= stridelen;
+    }
+}
+
+#define CPUID_UNKNOWN 2
+#define CPUID_AVAILABLE 0
+#define CPUID_UNAVAILABLE 1
+static int cpuid_state = 2;
+
+void aws_common_private_set_b64_simd_state(int state) {
```

Comment:
This is never used, and isn't documented either (how to use it, when to use it, what are valid parameter values).